### PR TITLE
SolidShapes and Visualizer

### PR DESCRIPTION
**Pull request by traversaro**
The changes are mainly : 

# [model] Added SolidShapes and related classes ( https://github.com/robotology/idyntree/commit/fbc5c622c4a947f179bfc0b2e0fce52690f9adb9 )

Useful for visualization and collision checking

# Added visualization library ( https://github.com/robotology/idyntree/commit/4336d0b7e52dee889bfcbd204afd8342a41597ed ) 

Add an optional depedency on the Irrlicht graphical engine,
choosen because it has a limited amount of dependency and it is
available on apt-get , homebrew and in binary form for Windows.

The dependency can be enabled with the IDYNTREE_USES_IRRLICHT variable.

The Visualizer class is currently quite limited, but the `idyntree-model-view`
has been added as a tool to showcase its use (for now it just visualize statically
a model in its default position).

The Visualizer run fine in `Matlab`. The only necessary workaround (at least on Ubuntu 16.04 and Matlab 2016b) was to avoid that Matlab loads its local version of the `libGL.so.1` file.
I managed to just rename it to avoid problems:
~~~
sudo mv /usr/local/MATLAB/R2016a/sys/opengl/lib/glnxa64/libGL.so.1 /usr/local/MATLAB/R2016a/sys/opengl/lib/glnxa64/libGL.so.1.back
~~~ 

After doing this, it is sufficient to run:
~~~
mdlLoader = iDynTree.ModelLoader();
mdlLoader.loadModelFromFile('model.urdf');

viz = iDynTree.Visualizer();
viz.addModel(mdlLoader.model(),'icub')
viz.draw();
~~~
To replicate the behavior of `idyntree-model-view`

### CODE REVIEW

File: src/model/include/iDynTree/Model/SolidShapes.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        /**
+         * Storage ot ModelSolidShapes.
+         */
+        std::vector< std::vector<SolidShape *> > linkSolidShapes;
+    };
+
+}
+
+#endif
```
Comment (francesco-romano) [inline src/model/include/iDynTree/Model/SolidShapes.h]: EOF

File: src/model/src/SolidShapes.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+
+    Sphere::~Sphere()
+    {
+    }
+
+    SolidShape* Sphere::clone()
+    {
+        return static_cast<SolidShape*>(new Sphere(*this));
```
Comment (francesco-romano) [inline src/model/src/SolidShapes.cpp]: is the static cast needed?

File: src/model/src/SolidShapes.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    ModelSolidShapes& ModelSolidShapes::operator=(const ModelSolidShapes& other)
+    {
+        return copy(other);
+    }
+
+    ModelSolidShapes::~ModelSolidShapes()
+    {
+        this->clear();
+    }
+}
```
Comment (francesco-romano) [inline src/model/src/SolidShapes.cpp]: EOF

File: src/model/src/SolidShapes.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        linkSolidShapes.resize(0);
+    }
+
+    ModelSolidShapes& ModelSolidShapes::copy(const ModelSolidShapes& other)
+    {
+        clear();
+        this->linkSolidShapes.resize(other.linkSolidShapes.size());
+        for(size_t link = 0; link < other.linkSolidShapes.size(); link++)
```
Comment (francesco-romano) [inline src/model/src/SolidShapes.cpp]: I love the fact that you started to remove spaces inside the brackets!!

To be meticulous, differently from a function call where there should not be space between the function name and the bracket, in `if`, `for`, etc there should be a space between them and the opening bracket 😝

File: src/model_io/urdf/include/iDynTree/ModelIO/URDFSolidShapesImport.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ */
+bool solidShapesFromURDFString(const std::string & urdf_string,
+                              const std::string & urdf_filename,
+                              const Model & model,
+                              const std::string urdfGeometryType,
+                                    iDynTree::ModelSolidShapes & output);
+
+}
+
+#endif
```
Comment (francesco-romano) [inline src/model_io/urdf/include/iDynTree/ModelIO/URDFSolidShapesImport.h]: EOF

File: src/tools/idyntree-model-view.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ * Authors: Silvio Traversaro
+ * CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ */
+
+#include <iDynTree/Model/Model.h>
+#include <iDynTree/ModelIO/ModelLoader.h>
+
+#include <iDynTree/Visualizer.h>
+
+#include "cmdline.h"
```
Comment (francesco-romano) [inline src/tools/idyntree-model-view.cpp]: What is `cmdline.h`? 
Is it internal to the executable sources?
If external, considering including it with `<>`

File: src/visualization/include/iDynTree/Visualizer.h [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    /**
+     * Close the visualizer.
+     */
+    void close();
+
+};
+
+}
+
+#endif
```
Comment (francesco-romano) [inline src/visualization/include/iDynTree/Visualizer.h]: EOF

File: src/visualization/src/IrrlichtUtils.h [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    if( !geomNode )
+    {
+        std::cerr << "[ERROR] addGeometryToSceneManager: geometry not loaded" << std::endl;
+        return 0;
+    }
+
+    // Set materials
+    if( use_iDynTree_material )
```
Comment (francesco-romano) [inline src/visualization/src/IrrlichtUtils.h]: Damn! I thought you learned to use spaces! 😝

File: src/visualization/src/IrrlichtUtils.h [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    matproj(0,0) *= -1;
+    camera->setProjectionMatrix(matproj);
+
+    return camera;
+}
+
+
+}
+
+#endif
```
Comment (francesco-romano) [inline src/visualization/src/IrrlichtUtils.h]: EOF

File: src/visualization/src/Visualizer.cpp [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+
+    pimpl->m_irrDevice->closeDevice();
+    pimpl->m_irrDevice->drop();
+    pimpl->m_irrDevice = 0;
+    pimpl->m_isInitialized = false;
+    return;
+#endif
+}
+
+
```
Comment (francesco-romano) [inline src/visualization/src/Visualizer.cpp]: You can add some more EOL here 😜

**Comment by francesco-romano** [comment]
Some inline comments on the various commits.

Additional comments:
1. (stupid) you can keep the last three commits in only one as all are related to CMake and warnings, aren't they?
2. Class hierarchy.
The problem is that you need a virtual method of the type:
`irr::scene::IMesh* createMesh()`, but it is highly tied to Irrlicht and has nothing to do with the Model library. Sometimes I really miss the Objective C class extensions 😢 

Anyway, I was thinking to this alternative, which is not very polymorphic (unfortunately) but if it makes sense to you will avoid all those methods:
You can create an enum `ShapeTypes`, and a virtual method `isOfShape(enum ShapeTypes)`.
All the classes implements this method returning `true` only to their type. Users can then be ''safely'' downcast to the respective type by using `static_cast`.

An additional alternative would be move the `dynamic_cast` in a wrapper class in the visualisation library, which then knows that it is using Irrlicht, thus you can wrap the `model::SolidShape` in this new class, and then simply call a `createMesh` function on the wrapped object.

File: src/model/src/SolidShapes.cpp (hunk 2 above)
Comment (traversaro) [inline src/model/src/SolidShapes.cpp]: I remember that some compiler had warnings related to implicit and C-style casting, but I could be wrong.

File: src/tools/idyntree-model-view.cpp (hunk 6 above)
Comment (traversaro) [inline src/tools/idyntree-model-view.cpp]: It is an header-only simple library to parse options from command line: https://github.com/robotology/idyntree/blob/master/src/tools/cmdline.h .

File: src/model/src/SolidShapes.cpp (hunk 2 above)
Comment (francesco-romano) [inline src/model/src/SolidShapes.cpp]: But in this case is from a child to parent. A cast should not be necessary, am I right?

Comment (traversaro) [inline src/model/src/SolidShapes.cpp]: I removed and no one complained. We can reinsert it when we enable the warnings.

File: src/model/src/SolidShapes.cpp (hunk 4 above)
Comment (traversaro) [inline src/model/src/SolidShapes.cpp]: I tried to fix them around, but it is a lost battle until we have some automatic check.

**Comment by traversaro** [comment]
> 1. (stupid) you can keep the last three commits in only one as all are related to CMake and warnings, aren't they?
Yes, but they are related to different components. I actually prefer "small" commits, such that is easier to track changes in history.

**Comment by traversaro** [comment]
I tried to fix all the style issues in https://github.com/robotology/idyntree/pull/215/commits/2195214c39dabf6fb46b969b91f05d73609360f8 , but as mentioned before it is a lost battle until we have some automatic warnings running.

**Comment by traversaro** [comment]
@claudiofantacci FYI, the script I mentioned today at lunch is the one added in this commit : https://github.com/robotology/idyntree/pull/215/commits/c2fce2333fa13c9a08fcf4bb145f05b17234ecb1 .

**Comment by claudiofantacci** [comment]
@traversaro thanks for the script!

**Pull request merged**